[Navigation] On intercepted navigations properly update url and history

TingPing opened the pull request:
#### 43c0f59f9fe5fa1c151c14f95132c90f075d4890
<pre>
[Navigation] On intercepted navigations properly update url and history
<a href="https://bugs.webkit.org/show_bug.cgi?id=276375">https://bugs.webkit.org/show_bug.cgi?id=276375</a>

Reviewed by Darin Adler.

Per the spec an interception can immediately apply the new navigation to
the browsers history.

<a href="https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigate-event-firing">https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigate-event-firing</a>:url-and-history-update-steps-2

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-replace-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-and-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-returns-non-promise-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-throws-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-multiple-times-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-popstate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-resolve-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-same-document-history-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-dynamic-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-history-push-same-url-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-intercept-history-state-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-state-repeated-await-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-state-repeated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-rejection-order-pagehide-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-replace-with-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-away-and-back-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateURLAndHistory):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/History.cpp:
(WebCore::History::stateObjectAdded):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::innerDispatchNavigateEvent):

Canonical link: <a href="https://commits.webkit.org/281063@main">https://commits.webkit.org/281063@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7e5b0d5b9ad6f3b560cc0f5c96f0185a63148acb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/58538 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/37865 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/11023 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/62164 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/8982 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/60667 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/45501 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/9179 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/47390 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/6396 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/60569 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/35443 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/50611 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/28242 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/32193 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/7913 "Found 1 new test failure: imported/w3c/web-platform-tests/mathml/relations/html5-tree/dynamic-childlist-002.html (failure)") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/7986 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/54145 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/8184 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/63867 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/2451 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/8192 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/54709 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/2459 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/50637 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/54788 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12935 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/2071 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/33694 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/34780 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/35864 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/34525 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1219,6 +1219,35 @@ static NavigationNavigationType determineNavigationType(FrameLoadType loadType,
+void FrameLoader::updateURLAndHistory(const URL& newURL, RefPtr<SerializedScriptValue>&& stateObject, NavigationHistoryBehavior historyHandling)
+{
+    RefPtr document = m_frame->document();
+    RefPtr documentLoader = protectedDocumentLoader();
```
This refs twice I think, common is to do:
RefPtr documentLoader = documentLoader();

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1219,6 +1219,35 @@ static NavigationNavigationType determineNavigationType(FrameLoadType loadType,
     return NavigationNavigationType::Push;
 }
 
+// https://html.spec.whatwg.org/multipage/browsing-the-web.html#url-and-history-update-steps
```
This needs a FIXME for integrating with History::stateObjectAdded.

TingPing replied:
I'm not sure I understand what you mean.

rwlbuis replied:
I mean that the code from History::stateObjectAdded is supposed to use the same algorithm:
https://html.spec.whatwg.org/multipage/nav-history-apis.html#shared-history-push/replace-state-steps

So basically, I think we want to avoid duplicating code that the code paths for history.pushState/replaceState and Navigation::innerDispatchNavigateEvent are doing, both ideally can just use the same algorithm (in FrameLoader::updateURLAndHistory).

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1219,6 +1219,35 @@ static NavigationNavigationType determineNavigationType(FrameLoadType loadType,
+
+    if (!stateObject)
+        stateObject = history->currentItem()->stateObject();
+
```
Needs refactoring now or later, i.e. next three lines are basically Document::updateURLForPushOrReplaceState.

rwlbuis commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1219,6 +1219,35 @@ static NavigationNavigationType determineNavigationType(FrameLoadType loadType,
+    else
+        history->pushState(WTFMove(stateObject), newURL.string());
+
+    // Consider the navigation completed.
```
Why are these two lines needed? AFAICS they are not listed as part of the spec algorithm.

TingPing replied:
This isn't the ideal place for it. I wanted to ensure the navigation was marked as completed but I'll keep that to a different change.

rwlbuis reviewed (commented):
Looks very promising, just a few questions and suggestions.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -613,6 +613,18 @@ void Navigation::abortOngoingNavigation(NavigateEvent& event)
+        return NavigationHistoryBehavior::Replace;
+    default:
+        return NavigationHistoryBehavior::Auto;
+    }
```
Here, you may want to include ASSERT_ NOT_REACHED logic, you can check how similar helper functions do that.

TingPing replied:
I feel it isn't necessary for one with a `default` personally.

rwlbuis commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -703,7 +715,8 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
             m_suppressNormalScrollRestorationDuringOngoingNavigation = true;
 
-        // FIXME: Step 7: URL and history update
+        if (navigationType == NavigationNavigationType::Push || navigationType == NavigationNavigationType::Replace)
```
navigationTypeToHistoryBehavior may be overkill, since we already narrowed it down to two NavigationNavigationTypes. I'd say we calculate NavigationHistoryBehavior in a local var for now, and if there are more users for it, then navigationTypeToHistoryBehavior starts making sense.

rwlbuis reviewed (commented):
This seems like a big improvement, LGTM. But I'd encourage to add a reminder/FIXME to not forget to refactor this more into existing FrameLoader/HistoryController code.

rwlbuis commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -352,6 +352,8 @@ class FrameLoader final : public CanMakeCheckedPtr<FrameLoader> {
     void loadItem(HistoryItem&, HistoryItem* fromItem, FrameLoadType, ShouldTreatAsContinuingLoad);
     HistoryItem* requestedHistoryItem() const { return m_requestedHistoryItem.get(); }
 
+    void updateURLAndHistory(const URL&, RefPtr<SerializedScriptValue>&& stateObject, NavigationHistoryBehavior historyHandling);
```
I think we need a default for historyHandling:
"an optional [history handling behavior](https://html.spec.whatwg.org/multipage/browsing-the-web.html#history-handling-behavior) historyHandling (default "[replace](https://html.spec.whatwg.org/multipage/browsing-the-web.html#navigationhistorybehavior-replace)"),"

darinadler replied:
Also could leave out the argument name since the type is unambiguous without a name.

darinadler commented on LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-and-navigate-expected.txt:
```diff
@@ -1,3 +1,3 @@
 
-FAIL Using intercept() then navigate() in the ensuing currententrychange should abort the finished promise (but not the committed promise) assert_equals: expected "#2" but got ""
+FAIL Using intercept() then navigate() in the ensuing currententrychange should abort the finished promise (but not the committed promise) assert_unreached: Should have rejected: undefined Reached unreachable code
```
Do you understand this failure?

rwlbuis replied:
I think I do, the test now gets past the location.hash assert line, but either we lack the logic to abort the back navigation because of the new navigation to "#2", or we do have the logic but it is not triggered (at the right time).  We do know some of our timing is off since we did not implement the "wait for all promises" logic yet, that could be related. Anyway, we'll revisit this in follow up PRs.

darinadler commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1219,6 +1219,32 @@ static NavigationNavigationType determineNavigationType(FrameLoadType loadType,
+{
+    ASSERT(m_frame->document());
+
+    if (documentLoader()->isInitialAboutBlank())
```
What guarantees `documentLoader()` is non-null?

rwlbuis replied:
The assumption I think is that the previous document exists and before committing to the new one we perform this algorithm so the document loader should be available. I added a check to the ASSERT since I think the (previous) document is in a strange state if there is no document loader available.

darinadler commented on Source/WebCore/loader/FrameLoader.cpp:
```diff
@@ -1219,6 +1219,32 @@ static NavigationNavigationType determineNavigationType(FrameLoadType loadType,
+
+    CheckedRef history = m_frame->checkedHistory();
+
+    // FIXME: Validate that stateObject deserializes without exception.
```
I don’t understand the meaning of the word “validate” here. Can we make it clearer what needs to be done?

rwlbuis replied:
I rephrased it, basically it is one of the algorithm steps.

darinadler commented on Source/WebCore/loader/FrameLoader.h:
```diff
@@ -352,6 +352,8 @@ class FrameLoader final : public CanMakeCheckedPtr<FrameLoader> {
     void loadItem(HistoryItem&, HistoryItem* fromItem, FrameLoadType, ShouldTreatAsContinuingLoad);
     HistoryItem* requestedHistoryItem() const { return m_requestedHistoryItem.get(); }
 
+    void updateURLAndHistory(const URL&, RefPtr<SerializedScriptValue>&& stateObject, NavigationHistoryBehavior historyHandling = NavigationHistoryBehavior::Replace);
```
Please remove the unnecessary argument name `historyHandling`.

rwlbuis replied:
Good point, I think the upload may have missed giving a warning about it. Fixed.

darinadler commented on Source/WebCore/page/Navigation.cpp:
```diff
@@ -706,7 +706,10 @@ bool Navigation::innerDispatchNavigateEvent(NavigationNavigationType navigationT
 
-        // FIXME: Step 7: URL and history update
+        if (navigationType == NavigationNavigationType::Push || navigationType == NavigationNavigationType::Replace) {
+            auto historyHandling = navigationType == NavigationNavigationType::Replace ? NavigationHistoryBehavior::Replace : NavigationHistoryBehavior::Push;
```
Awkward that we need to translate `Push` to `Push` and `Replace` to `Replace`. Wish there was a more elegant way to write this.

rwlbuis replied:
Yes, and also StateObjectType enum by now seems superfluous. So, if Navigation API becomes more accepted, I guess it will make sense to use more of its enums and constants, like I think happened for Fetch API in the past.

TingPing committed:
[Navigation] On intercepted navigations properly update url and history
https://bugs.webkit.org/show_bug.cgi?id=276375

Reviewed by Darin Adler.

Per the spec an interception can immediately apply the new navigation to
the browsers history.

https://html.spec.whatwg.org/multipage/nav-history-apis.html#navigate-event-firing:url-and-history-update-steps-2

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/currententrychange-event/navigation-navigate-replace-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-and-navigate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-returns-non-promise-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-throws-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-history-pushState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-history-replaceState-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-multiple-times-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-popstate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-resolve-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-same-document-history-back-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/signal-abort-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-activation/activation-replace-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-history-entry/no-referrer-dynamic-url-censored-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-history-push-same-url-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-intercept-history-state-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-state-repeated-await-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/navigate-state-repeated-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-interrupted-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-intercept-rejected-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-rejection-order-pagehide-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/per-entry-events/dispose-same-document-replace-with-intercept-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/state/cross-document-away-and-back-expected.txt:
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::updateURLAndHistory):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/page/History.cpp:
(WebCore::History::stateObjectAdded):
* Source/WebCore/page/Navigation.cpp:
(WebCore::Navigation::innerDispatchNavigateEvent):

Canonical link: https://commits.webkit.org/281063@main
